### PR TITLE
L2 shorts - Move loans state out of collateral state 

### PR DESCRIPTION
**Pull request by jacko125**

### CODE REVIEW

File: contracts/Collateral.sol [hunk 1]
```diff
@@ -129,60 +125,24 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
-        uint cvalue = _exchangeRates().effectiveValue(collateralKey, loan.collateral, sUSD);
-        uint dvalue = _exchangeRates().effectiveValue(loan.currency, loan.amount.add(loan.accruedInterest), sUSD);
-        cratio = cvalue.divideDecimal(dvalue);
+    function collateralRatio(uint id) public view returns (uint cratio) {
+        Loan memory loan = loans[id];
+        return _collateralUtil().getCollateralRatio(loan, collateralKey);
     }
 
     // The maximum number of synths issuable for this amount of collateral
     function maxLoan(uint amount, bytes32 currency) public view returns (uint max) {
```
Comment (jacko125) [inline contracts/Collateral.sol:144]: potentially move all to CollateralUtil for dapp support if contract size limit hit

File: contracts/Collateral.sol [hunk 2]
```diff
@@ -291,27 +230,24 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
-
-        // 4. Cap the number of loans so that the array doesn't get too big.
-        require(state.getNumLoans(msg.sender) < maxLoansPerAccount, "Max loans exceeded");
+        require(collateral >= minCollateral);
 
         // 5. Check we haven't hit the debt cap for non snx collateral.
-        (bool canIssue, bool anyRateIsInvalid) = _manager().exceedsDebtLimit(amount, currency);
+        (bool canIssue, bool anyRateIsInvalid) = manager.exceedsDebtLimit(amount, currency);
 
-        require(canIssue && !anyRateIsInvalid, "Debt limit or invalid rate");
```
Comment (jacko125) [inline contracts/Collateral.sol:311]: Add these require reasons back in if size permits

File: publish/deployed/local-ovm/params.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+		MAX_DEBT: w3utils.toWei('75000000'), // 75 million sUSD
+		BASE_BORROW_RATE: Math.round((0.005 * 1e18) / 31556926).toString(), // 31556926 is CollateralManager seconds per year
+		BASE_SHORT_RATE: Math.round((0.005 * 1e18) / 31556926).toString(),
+	},
+	COLLATERAL_SHORT: {
+		SYNTHS: ['sETH'],
+		MIN_CRATIO: w3utils.toWei('1.2'),
+		MIN_COLLATERAL: w3utils.toWei('1000'), // 1000 sUSD
+		ISSUE_FEE_RATE: w3utils.toWei('0.005'),
+		INTERACTION_DELAY: '0', // 0 secs
```
Comment (gberg1) [inline publish/deployed/local-ovm/params.js:18]: Confirm with the council what they want for min interaction delay given web sockets

File: contracts/Collateral.sol [hunk 4]
```diff
@@ -420,140 +309,106 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
         address liquidator,
-        Loan memory loan
-    ) internal returns (uint collateral) {
+        Loan storage loan
+    ) internal returns (uint amount, uint collateral) {
+        (amount, collateral) = _closeLoan(borrower, liquidator, loan);
+
+        // 7. Record loan as closed
+        // TODO: why is this commented out? Want to allow the same loan to be used when liquidated?
+        // loan.interestIndex = 0;
```
Comment (jacko125) [inline contracts/Collateral.sol:318]: Uncomment this as per TODO

File: contracts/Collateral.sol [hunk 5]
```diff
@@ -420,140 +309,106 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
         } else {
-            _manager().decrementLongs(loan.currency, loan.amount);
+            manager.decrementLongs(loan.currency, loan.amount);
         }
 
         // 6. Pay fees
         _payFees(loan.accruedInterest, loan.currency);
-
-        // 7. Record loan as closed
-        loan.amount = 0;
```
Comment (jacko125) [inline contracts/Collateral.sol:453]: Should loan amounts be zeroed out after closing. 

Loans that have been closed should they be re-usable for depositing more collateral / drawing down more ?

File: contracts/Collateral.sol [hunk 6]
```diff
@@ -420,140 +309,106 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
-        loan.collateral = 0;
-        loan.accruedInterest = 0;
-        loan.interestIndex = 0;
```
Comment (jacko125) [inline contracts/Collateral.sol:456]: interestIndex = 0 is used to indicate loan is closed. 

- deposit collateral  checks interestIndex != 0
- draw down of collateral checks interestIndex != 0

File: contracts/Collateral.sol [hunk 7]
```diff
@@ -563,26 +418,20 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
         // 12. Work out how much collateral to redeem.
-        collateralLiquidated = collateralRedeemed(loan.currency, amountToLiquidate);
+        collateralLiquidated = _collateralUtil().collateralRedeemed(loan.currency, amountToLiquidate, collateralKey);
         loan.collateral = loan.collateral.sub(collateralLiquidated);
 
-        // 13. Update the last interaction time.
-        loan.lastInteraction = block.timestamp;
-
         // 14. Burn the synths from the liquidator.
-        require(!_exchanger().hasWaitingPeriodOrSettlementOwing(msg.sender, loan.currency), "Waiting or settlement owing");
```
Comment (jacko125) [inline contracts/Collateral.sol:580]: only valid for L2 where the fee reclamation is default to 0. 

On L1 this assumption would be incorrect for `Collateral.sol`

File: contracts/Collateral.sol [hunk 8]
```diff
@@ -592,60 +441,41 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
         // 3. Check loan is open and last interaction time.
-        _checkLoanAvailable(loan);
-
-        // 4. Accrue interest.
-        loan = accrueInterest(loan);
-
-        // 5. Check the spender has enough synths to make the repayment
-        _checkSynthBalance(repayer, loan.currency, payment);
+        // TODO: why is there an interaction delay required for repaying?
+        //   The user already owns the eth and is burning it, how do frontrunners win here?
```
Comment (jacko125) [inline contracts/Collateral.sol:449]: Stops user closing the short or loan right after opening without waiting for interaction delay to take advantage of price volatility.

File: contracts/Collateral.sol [hunk 9]
```diff
@@ -592,60 +441,41 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
         // 6. Process the payment.
-        loan = _processPayment(loan, payment);
+        require(payment > 0);
+        _processPayment(loan, payment);
 
         // 7. Update the last interaction time.
         loan.lastInteraction = block.timestamp;
 
         // 8. Burn synths from the payer
-        require(!_exchanger().hasWaitingPeriodOrSettlementOwing(repayer, loan.currency), "Waiting or settlement owing");
```
Comment (jacko125) [inline contracts/Collateral.sol:621]: assumption that fee reclamation could possibly still be here. On L1 this is also the case.

File: test/contracts/setup.js [hunk 10]
```diff
@@ -244,6 +244,16 @@ const setupContract = async ({
...
+		CollateralUtil: [tryGetAddressOf('AddressResolver')],
+		Collateral: [
+			tryGetAddressOf('CollateralState'),
+			tryGetAddressOf('CollateralManager'),
+			tryGetAddressOf('AddressResolver'),
+			'sUSD',
+			1.2,
+			100,
+		],
+		CollateralState: [owner, tryGetAddressOf('Collateral')],
```
Comment (jacko125) [inline test/contracts/setup.js:256]: let's remove CollateralState from this branch as won't be needed.

File: contracts/SystemSettings.sol [hunk 11]
```diff
@@ -271,6 +271,43 @@ contract SystemSettings is Owned, MixinSystemSettings, ISystemSettings {
         }
     }
 
+    function setMinCratio(uint minCratio, address collateral) external onlyOwner {
```
Comment (jacko125) [inline contracts/SystemSettings.sol:274]: update function argument for setters

File: contracts/Collateral.sol [hunk 12]
```diff
@@ -60,39 +59,36 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
         bytes32 _collateralKey,
         uint _minCratio,
         uint _minCollateral
-    ) public Owned(_owner) MixinSystemSettings(_resolver) {
+    ) public onlyOwner {
+        require(!initialized);
         manager = _manager;
-        state = _state;
         collateralKey = _collateralKey;
         minCratio = _minCratio;
```
Comment (jacko125) [inline contracts/Collateral.sol:89]: This should be now set on the SystemSettings as part of deploy script and removed from the `initialize` step.

File: contracts/Collateral.sol [hunk 13]
```diff
@@ -35,10 +34,10 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
     // The synth corresponding to the collateral.
     bytes32 public collateralKey;
 
-    // Stores loans
-    CollateralState public state;
+    // Stores loans open
+    mapping(uint => Loan) public loans;
```
Comment (jacko125) [inline contracts/Collateral.sol:38]: All the loans can be internalised inside each Collateral contract, which is now managed as a mapping and removed CollateralState contract.

Reduces the complexity of storing the loans in an external state and each Collateral contract isn't upgradable as well so the external state pattern wasn't needed.

File: contracts/Collateral.sol [hunk 14]
```diff
@@ -60,39 +59,36 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
     // The fee charged for issuing a loan.
     uint public issueFeeRate;
 
-    // The maximum number of loans that an account can create with this collateral.
-    uint public maxLoansPerAccount = 50;
+    bool public canOpenLoans = true;
```
Comment (jacko125) [inline contracts/Collateral.sol:62]: move `canOpenLoans` check to SystemSettings

File: contracts/Collateral.sol [hunk 15]
```diff
@@ -265,20 +204,20 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
+    function setManager(ICollateralManager _newManager) external onlyOwner {
         manager = _newManager;
         emit ManagerUpdated(manager);
     }
 
-    function setCanOpenLoans(bool _canOpenLoans) external onlyOwner {
-        canOpenLoans = _canOpenLoans;
-        emit CanOpenLoansUpdated(canOpenLoans);
-    }
+    // function setCanOpenLoans(bool _canOpenLoans) external onlyOwner {
```
Comment (jacko125) [inline contracts/Collateral.sol:217]: Can remove the commented code out

File: contracts/Collateral.sol [hunk 16]
```diff
@@ -291,27 +230,24 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
 
         // 5. Check we haven't hit the debt cap for non snx collateral.
-        (bool canIssue, bool anyRateIsInvalid) = _manager().exceedsDebtLimit(amount, currency);
+        (bool canIssue, bool anyRateIsInvalid) = manager.exceedsDebtLimit(amount, currency);
 
-        require(canIssue && !anyRateIsInvalid, "Debt limit or invalid rate");
+        require(canIssue && !anyRateIsInvalid);
 
         // 6. Require requested loan < max loan
-        require(amount <= maxLoan(collateral, currency), "Exceeds max borrowing power");
```
Comment (jacko125) [inline contracts/Collateral.sol:314]: require reason removed to reduce contract size - Add these require reasons back if it fits

File: contracts/Collateral.sol [hunk 17]
```diff
@@ -320,98 +256,51 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
-        // 9. Pay fees
-        _payFees(loan.accruedInterest, loan.currency);
-
-        // 10. Record loan as closed
-        loan.amount = 0;
-        loan.collateral = 0;
-        loan.accruedInterest = 0;
+        // 10. Record loan as closed.
         loan.interestIndex = 0;
-        loan.lastInteraction = block.timestamp;
```
Comment (jacko125) [inline contracts/Collateral.sol:413]: We use loan.interestIndex to set the loan as closed. Not needed to set `loan.lastInteraction` as a closed loan isn't able to be used again.

File: contracts/Collateral.sol [hunk 18]
```diff
@@ -672,74 +497,47 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
-
-        require(!invalid, "Rates are invalid");
-
-        // 3. Get the time since we last updated the rate.
-        uint timeDelta = block.timestamp.sub(lastUpdated).mul(SafeDecimalMath.unit());
-
-        // 4. Get the latest cumulative rate. F_n+1 = F_n + F_last
-        uint latestCumulative = lastRate.add(rate.multiplyDecimal(timeDelta));
+    function accrueInterest(Loan storage loan) internal {
+        (uint differential, uint newIndex) = manager.accrueInterest(loan.interestIndex, loan.currency, loan.short);
```
Comment (jacko125) [inline contracts/Collateral.sol:506]: accruing interest is done on the CollateralManager now

File: contracts/Collateral.sol [hunk 19]
```diff
@@ -755,6 +553,21 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
         }
     }
 
+    function _getLoanAndAccrueInterest(uint id, address owner) internal returns (Loan storage loan) {
```
Comment (jacko125) [inline contracts/Collateral.sol:554]: internal function to retrieve the loan and accrueInterest each time a loan is intereacted with, whether its drawing down, repaying or liquidating.

File: contracts/CollateralErc20.sol [hunk 20]
```diff
@@ -18,107 +17,107 @@ contract CollateralErc20 is ICollateralErc20, Collateral {
...
         underlyingContractDecimals = _underlyingDecimals;
     }
 
-    function open(
-        uint collateral,
-        uint amount,
-        bytes32 currency
-    ) external {
-        require(collateral <= IERC20(underlyingContract).allowance(msg.sender, address(this)), "Allowance not high enough");
+    // function open(
```
Comment (jacko125) [inline contracts/CollateralErc20.sol:35]: @gberg1 @justinjmoses 

All the functions in `CollateralErc20.sol` and `CollateralEth.sol` should be uncommented. We already exclude the compiling of these contracts in `ovm-ignore.json` and  deployment to L2.

File: contracts/CollateralEth.sol [hunk 21]
```diff
@@ -7,70 +7,68 @@ import "./Collateral.sol";
...
-
-    function open(uint amount, bytes32 currency) external payable {
-        openInternal(msg.value, amount, currency, false);
+    ) public Collateral(_owner, _resolver) {
+        initialize(_manager, _collateralKey, _minCratio, _minCollateral);
     }
 
-    function close(uint id) external {
-        uint collateral = closeInternal(msg.sender, id);
+    // function open(uint amount, bytes32 currency) external payable {
```
Comment (jacko125) [inline contracts/CollateralEth.sol:25]: Same here. All the functions in CollateralEth.sol should be uncommented. They were commented out by Mike to get CollateralShorts compiling on OVM.

We already exclude the compiling of these contracts in ovm-ignore.json and deployment to L2.

File: contracts/CollateralManager.sol [hunk 22]
```diff
@@ -425,10 +419,10 @@ contract CollateralManager is ICollateralManager, Owned, Pausable, MixinResolver
 
         // first check contract state
         for (uint i = 0; i < requiredSynthNamesInResolver.length; i++) {
-            bytes32 synthName = requiredSynthNamesInResolver[i];
-            if (!_shortableSynths.contains(synthName) || synthToInverseSynth[synthName] == bytes32(0)) {
-                return false;
-            }
+            // bytes32 synthName = requiredSynthNamesInResolver[i];
+            // if (!_shortableSynths.contains(synthName) || synthToInverseSynth[synthName] == bytes32(0)) {
```
Comment (jacko125) [inline contracts/CollateralManager.sol:423]: This commented section is addressed in BaseCollateralManager https://github.com/Synthetixio/synthetix/blob/415a83f7e957313c22079d843067c08851ff84f7/contracts/BaseCollateralManager.sol#L417-L444

We only need to check that the `_shortableSynths` is set as `synthToInverseSynth` should be set as well within the `CollateralManagerWithInverse` contract.

File: contracts/CollateralManager.sol [hunk 23]
```diff
@@ -452,7 +446,7 @@ contract CollateralManager is ICollateralManager, Owned, Pausable, MixinResolver
                 state.removeShortCurrency(synthKey);
 
                 // remove the inverse mapping.
-                delete synthToInverseSynth[synths[i]];
+                // delete synthToInverseSynth[synths[i]];
```
Comment (jacko125) [inline contracts/CollateralManager.sol:449]: Subclass of collateralManager manages the logic here to remove synths / inverse synths.

https://github.com/Synthetixio/synthetix/blob/415a83f7e957313c22079d843067c08851ff84f7/contracts/BaseCollateralManager.sol#L447

File: contracts/CollateralState.sol [hunk 24]
```diff
@@ -24,6 +24,7 @@ contract CollateralState is Owned, State, ICollateralLoan {
         Loan[] memory accountLoans = loans[account];
```
Comment (jacko125) [inline contracts/CollateralState.sol:24]: CollateralState contract isn't required anymore for the refactored `Collateral.sol` contract. 

Existing unit tests. / prod tests for Multi-collateral loans should be updated to remove setup of the collateralState contract.

File: contracts/ShortingRewards.sol [hunk 25]
```diff
@@ -112,7 +112,7 @@ contract ShortingRewards is IShortingRewards, RewardsDistributionRecipient, Reen
     }
 
     // this can be called by anyone on the short contract to claim the rewards.
-    function getReward(address account) external onlyShortContract nonReentrant updateReward(account) {
+    function getReward(address account) external nonReentrant updateReward(account) {
```
Comment (jacko125) [inline contracts/ShortingRewards.sol:115]: Shorting rewards can be claimed directly on the reward contract instead of proxying through the shorts contract.

This removes extra function in `CollateralShort` to `getRewards` which reduces the deploy contract size for `CollateralShort`.

File: contracts/CollateralShort.sol [hunk 26]
```diff
@@ -79,10 +64,4 @@ contract CollateralShort is Collateral {
 
         IERC20(address(_synthsUSD())).transfer(msg.sender, collateralLiquidated);
     }
-
-    function getReward(bytes32 currency, address account) external {
```
Comment (jacko125) [inline contracts/CollateralShort.sol:83]: Shorting rewards can be claimed directly on the reward contract instead of proxying through the shorts contract.

This removes extra function in CollateralShort to getRewards which reduces the deploy contract size for CollateralShort.

File: contracts/Collateral.sol [hunk 27]
```diff
@@ -129,60 +125,24 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
         return IExchangeRates(requireAndGetAddress(CONTRACT_EXRATES));
     }
 
-    function _exchanger() internal view returns (IExchanger) {
```
Comment (jacko125) [inline contracts/Collateral.sol:132]: We need to add `_exchanger()` back if we are checking whether the fee reclamation period is over before burning synths to repay a loan.

File: contracts/Collateral.sol [hunk 28]
```diff
@@ -420,140 +307,101 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
 
         // 2. Store this for the event.
-        uint amount = loan.amount;
+        amount = loan.amount;
 
         // 3. Return collateral to the child class so it knows how much to transfer.
         collateral = loan.collateral;
 
         // 4. Burn the synths
-        require(!_exchanger().hasWaitingPeriodOrSettlementOwing(liquidator, loan.currency), "Waiting or settlement owing");
```
Comment (jacko125) [inline contracts/Collateral.sol:435]: Should keep this check as assumption that fee reclamation could possibly still be here on L2 and is only set to 0 currently. On L1 this is also the case where fee reclamation still needs to be checked.

File: contracts/CollateralErc20.sol [hunk 29]
```diff
@@ -7,7 +7,6 @@ import "./Collateral.sol";
 import "./interfaces/ICollateralErc20.sol";
 
 // Internal references
-import "./CollateralState.sol";
```
Comment (gberg1) [inline contracts/CollateralErc20.sol:10]: Strategy: Move CollateralState.sol into an internal mapping

File: contracts/CollateralErc20.sol [hunk 30]
```diff
@@ -18,107 +17,107 @@ contract CollateralErc20 is ICollateralErc20, Collateral {
     uint public underlyingContractDecimals;
 
     constructor(
-        CollateralState _state,
         address _owner,
-        address _manager,
+        ICollateralManager _manager,
```
Comment (gberg1) [inline contracts/CollateralErc20.sol:21]: Strategy 2: Make CollateralManager an interface

File: contracts/CollateralErc20.sol [hunk 31]
```diff
@@ -18,107 +17,107 @@ contract CollateralErc20 is ICollateralErc20, Collateral {
...
         address _resolver,
         bytes32 _collateralKey,
         uint _minCratio,
         uint _minCollateral,
         address _underlyingContract,
         uint _underlyingDecimals
-    ) public Collateral(_state, _owner, _manager, _resolver, _collateralKey, _minCratio, _minCollateral) {
-        underlyingContract = _underlyingContract;
+    ) public Collateral(_owner, _resolver) {
+        initialize(_manager, _collateralKey, _minCratio, _minCollateral);
```
Comment (gberg1) [inline contracts/CollateralErc20.sol:29]: Strategy 3: Change constructor to use initialize pattern

File: contracts/CollateralEth.sol [hunk 32]
```diff
@@ -7,70 +7,68 @@ import "./Collateral.sol";
...
-
 // This contract handles the payable aspects of eth loans.
 contract CollateralEth is Collateral, ICollateralEth, ReentrancyGuard {
     mapping(address => uint) public pendingWithdrawals;
 
     constructor(
-        CollateralState _state,
         address _owner,
-        address _manager,
+        ICollateralManager _manager,
```
Comment (gberg1) [inline contracts/CollateralEth.sol:16]: Strategy 2: Use interface for CollateralManager

File: contracts/CollateralEth.sol [hunk 33]
```diff
@@ -7,70 +7,68 @@ import "./Collateral.sol";
...
         address _resolver,
         bytes32 _collateralKey,
         uint _minCratio,
         uint _minCollateral
-    ) public Collateral(_state, _owner, _manager, _resolver, _collateralKey, _minCratio, _minCollateral) {}
-
-    function open(uint amount, bytes32 currency) external payable {
-        openInternal(msg.value, amount, currency, false);
+    ) public Collateral(_owner, _resolver) {
```
Comment (gberg1) [inline contracts/CollateralEth.sol:21]: Strategy 3: Initialize for constructor

File: contracts/CollateralManager.sol [hunk 34]
```diff
@@ -201,7 +196,8 @@ contract CollateralManager is ICollateralManager, Owned, Pausable, MixinResolver
         }
     }
 
-    function getBorrowRate() external view returns (uint borrowRate, bool anyRateIsInvalid) {
+    // TODO: Enforce a maximum?
+    function getBorrowRate() public view returns (uint borrowRate, bool anyRateIsInvalid) {
```
Comment (gberg1) [inline contracts/CollateralManager.sol:200]: Strategy 4: Change external to public funcs in CollateralManager

File: contracts/CollateralManager.sol [hunk 35]
```diff
@@ -223,16 +219,14 @@ contract CollateralManager is ICollateralManager, Owned, Pausable, MixinResolver
         anyRateIsInvalid = ratesInvalid;
     }
 
-    function getShortRate(bytes32 synth) external view returns (uint shortRate, bool rateIsInvalid) {
-        bytes32 synthKey = _synth(synth).currencyKey();
-
+    // TODO: Enforce a maximum?
+    function getShortRate(bytes32 synthKey) public view returns (uint shortRate, bool rateIsInvalid) {
```
Comment (gberg1) [inline contracts/CollateralManager.sol:223]: Strategy 4: Change external to public func in collateral manager

File: contracts/CollateralManager.sol [hunk 36]
```diff
@@ -223,16 +219,14 @@ contract CollateralManager is ICollateralManager, Owned, Pausable, MixinResolver
...
         rateIsInvalid = _exchangeRates().rateIsInvalid(synthKey);
 
-        // get the spot supply of the synth, its iSynth
-        uint longSupply = IERC20(address(_synth(synth))).totalSupply();
-        uint inverseSupply = IERC20(address(_synth(synthToInverseSynth[synth]))).totalSupply();
+        // get the long supply of
```
Comment (gberg1) [inline contracts/CollateralManager.sol:226]: Strategy 5: remove getting the isynths. Why is this needed again?

File: contracts/CollateralManager.sol [hunk 37]
```diff
@@ -461,11 +455,19 @@ contract CollateralManager is ICollateralManager, Owned, Pausable, MixinResolver
 
     /* ---------- STATE MUTATIONS ---------- */
 
-    function updateBorrowRates(uint rate) external onlyCollateral {
+    function updateBorrowRates(uint rate) internal {
```
Comment (gberg1) [inline contracts/CollateralManager.sol:458]: Strategy 1: changes collateralstate to use an internal mapping to store the state

File: contracts/CollateralShort.sol [hunk 38]
```diff
@@ -5,37 +5,22 @@ pragma experimental ABIEncoderV2;
...
-        uint _minCollateral
-    ) public Collateral(_state, _owner, _manager, _resolver, _collateralKey, _minCratio, _minCollateral) {}
+    constructor(address _owner, address _resolver) public Collateral(_owner, _resolver) {}
 
     function open(
         uint collateral,
         uint amount,
         bytes32 currency
-    ) external {
-        require(
```
Comment (gberg1) [inline contracts/CollateralShort.sol:27]: Why did we remove this check for allowance?

File: contracts/CollateralUtil.sol [hunk 39]
```diff
@@ -0,0 +1,85 @@
+pragma solidity ^0.5.16;
```
Comment (gberg1) [inline contracts/CollateralUtil.sol:1]: Strategy 6: Move the view funcs from Collateral.sol to a utility contract

File: contracts/MixinSystemSettings.sol [hunk 40]
```diff
@@ -30,6 +30,10 @@ contract MixinSystemSettings is MixinResolver {
     bytes32 internal constant SETTING_ETHER_WRAPPER_MAX_ETH = "etherWrapperMaxETH";
     bytes32 internal constant SETTING_ETHER_WRAPPER_MINT_FEE_RATE = "etherWrapperMintFeeRate";
     bytes32 internal constant SETTING_ETHER_WRAPPER_BURN_FEE_RATE = "etherWrapperBurnFeeRate";
+    bytes32 internal constant SETTING_MIN_CRATIO = "minCratio";
```
Comment (gberg1) [inline contracts/MixinSystemSettings.sol:33]: Strategy 7: Move the settings from Collateral.sol into SystemSettings / MixinSystemSettings taking into account variable config

File: contracts/Collateral.sol (hunk 27 above)
Comment (gberg1) [inline contracts/Collateral.sol:132]: Strategy 8: Removed exchanger from Collateral.sol

File: contracts/Collateral.sol [hunk 41]
```diff
@@ -207,33 +167,12 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
         return true;
     }
 
-    /* ---------- UTILITIES ---------- */
```
Comment (gberg1) [inline contracts/Collateral.sol:210]: Strategy 9: Move the Utilities out of Collateral.sol. Where did these get moved to?

File: contracts/Collateral.sol [hunk 42]
```diff
@@ -207,33 +167,12 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
-        ratio = SafeDecimalMath.unit().divideDecimalRound(minCratio);
-    }
-
     /* ========== MUTATIVE FUNCTIONS ========== */
 
     /* ---------- Synths ---------- */
 
     function addSynths(bytes32[] calldata _synthNamesInResolver, bytes32[] calldata _synthKeys) external onlyOwner {
-        require(_synthNamesInResolver.length == _synthKeys.length, "Input array length mismatch");
+        require(_synthNamesInResolver.length == _synthKeys.length);
```
Comment (gberg1) [inline contracts/Collateral.sol:175]: Strategy 10: Remove the error messages in require statements. This isn't ideal need to think through how we want to manage the messages thrown when a require statement throws

File: contracts/Collateral.sol [hunk 43]
```diff
@@ -291,27 +230,24 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
+        require(synthsByKey[currency] > 0);
 
         // 2. Make sure the synth rate is not invalid.
-        require(!_exchangeRates().rateIsInvalid(currency), "Currency rate is invalid");
+        require(!_exchangeRates().rateIsInvalid(currency));
 
         // 3. Collateral >= minimum collateral size.
-        require(collateral >= minCollateral, "Not enough collateral to open");
-
-        // 4. Cap the number of loans so that the array doesn't get too big.
```
Comment (gberg1) [inline contracts/Collateral.sol:305]: We're removing the cap to the number of loans?

File: contracts/Collateral.sol [hunk 44]
```diff
@@ -320,98 +256,49 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
-                amount: amount,
-                short: short,
-                accruedInterest: 0,
-                interestIndex: 0,
-                lastInteraction: block.timestamp
-            });
+        id = manager.getNewLoanId();
+
+        // 10. Create the loan.
+        loans[id] = Loan({
```
Comment (gberg1) [inline contracts/Collateral.sol:262]: Strategy 11: Move the Loan memory to loans

File: contracts/Collateral.sol [hunk 45]
```diff
@@ -320,98 +256,49 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
+            accruedInterest: 0,
+            interestIndex: 0,
+            lastInteraction: block.timestamp
+        });
 
         // 11. Accrue interest on the loan.
-        loan = accrueInterest(loan);
+        accrueInterest(loans[id]);
 
-        // 12. Save the loan to storage
```
Comment (gberg1) [inline contracts/Collateral.sol:342]: No longer need to save loan to storage

File: contracts/Collateral.sol [hunk 46]
```diff
@@ -420,140 +307,101 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
-
-        // 2. Get the loan
-        Loan memory loan = state.getLoan(account, id);
-
-        // 3. Check loan is open and last interaction time.
-        _checkLoanAvailable(loan);
-
-        // 4. Accrue interest
-        loan = accrueInterest(loan);
+        Loan storage loan = loans[id];
```
Comment (gberg1) [inline contracts/Collateral.sol:360]: Don't have context on this change, need to review this

File: contracts/Collateral.sol [hunk 47]
```diff
@@ -420,140 +307,101 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
-    function withdrawInternal(uint id, uint amount) internal rateIsValid returns (uint withdraw) {
-        // 0. Check the system is active.
-        _systemStatus().requireIssuanceActive();
 
-        // 1. Get the loan.
-        Loan memory loan = state.getLoan(msg.sender, id);
-
-        // 2. Check loan is open and last interaction time.
-        _checkLoanAvailable(loan);
+        return (loan.amount, loan.collateral);
```
Comment (gberg1) [inline contracts/Collateral.sol:371]: Strategy 12: Return the amount and collateral from the loan

File: contracts/Collateral.sol [hunk 48]
```diff
@@ -563,26 +411,22 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
 
         // 10. If its greater than the amount owing, we need to close the loan.
         if (amountToLiquidate >= amountOwing) {
-            return closeByLiquidationInternal(borrower, msg.sender, loan);
+            (, collateralLiquidated) = closeByLiquidationInternal(borrower, msg.sender, loan);
+            return collateralLiquidated;
         }
 
+        // require(IERC20(address(_synth(synthsByKey[loan.currency]))).balanceOf(msg.sender) >= amountToLiquidate);
```
Comment (gberg1) [inline contracts/Collateral.sol:418]: What is this commented out require doing?

File: contracts/Collateral.sol [hunk 49]
```diff
@@ -563,26 +411,22 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
         // 11. Process the payment to workout interest/principal split.
-        loan = _processPayment(loan, amountToLiquidate);
+        _processPayment(loan, amountToLiquidate);
 
         // 12. Work out how much collateral to redeem.
-        collateralLiquidated = collateralRedeemed(loan.currency, amountToLiquidate);
+        collateralLiquidated = _collateralUtil().collateralRedeemed(loan.currency, amountToLiquidate, collateralKey);
         loan.collateral = loan.collateral.sub(collateralLiquidated);
 
-        // 13. Update the last interaction time.
```
Comment (gberg1) [inline contracts/Collateral.sol:576]: Same question above. removing this interaction time update under the assumption that the interaction delay will be 0?

File: contracts/Collateral.sol [hunk 50]
```diff
@@ -592,60 +436,39 @@ contract Collateral is ICollateralLoan, Owned, MixinSystemSettings {
...
 
         // 3. Check loan is open and last interaction time.
-        _checkLoanAvailable(loan);
-
-        // 4. Accrue interest.
-        loan = accrueInterest(loan);
-
-        // 5. Check the spender has enough synths to make the repayment
-        _checkSynthBalance(repayer, loan.currency, payment);
+        require(loan.lastInteraction.add(interactionDelay) <= block.timestamp);
```
Comment (gberg1) [inline contracts/Collateral.sol:443]: How will we deal with the potential worst case of for 5 minutes the Block.timestamp not changing on L2? This seems like an issue that we can't use block.timestamp here? @justinjmoses

File: contracts/CollateralManager.sol [hunk 51]
```diff
@@ -485,6 +487,35 @@ contract CollateralManager is ICollateralManager, Owned, Pausable, MixinResolver
         state.decrementShorts(synth, amount);
     }
 
+    function accrueInterest(
```
Comment (gberg1) [inline contracts/CollateralManager.sol:490]: Strategy 15: Move accrueInterest to CollateralManager

File: contracts/CollateralManager.sol [hunk 52]
```diff
@@ -223,16 +219,14 @@ contract CollateralManager is ICollateralManager, Owned, Pausable, MixinResolver
...
         rateIsInvalid = _exchangeRates().rateIsInvalid(synthKey);
 
-        // get the spot supply of the synth, its iSynth
-        uint longSupply = IERC20(address(_synth(synth))).totalSupply();
-        uint inverseSupply = IERC20(address(_synth(synthToInverseSynth[synth]))).totalSupply();
+        // get the long supply of
+        uint longSupply = IERC20(address(_synth(shortableSynthsByKey[synthKey]))).totalSupply();
         // add the iSynth to supply properly reflect the market skew.
-        uint shortSupply = state.short(synthKey).add(inverseSupply);
+        uint shortSupply = state.short(synthKey);
```
Comment (jacko125) [inline contracts/CollateralManager.sol:229]: Add a way for the CollateralManager to check dynamic list of ETH / BTC / LINK / other collateral that could offset the skew for these assets being shorted.

**Comment by barrasso** [comment]
Duplicate: https://github.com/Synthetixio/synthetix/pull/1448
